app/views/admins/kudos/index.html.erb

```diff
@@ -0,0 +1,23 @@
+<h1>Kudos listing page</h1>
+<table>
+  <thead>
+    <tr>
+      <th>Title</th>
+      <th>Content</th>
+      <th>Given by</th>
+      <th>Received by</th>
+      <th colspan ="2">Delete Kudo</th>
+    </tr>
+  </thead>
+
+  <tbody>
+    <% @kudos.each do |kudo| %>
+      <tr>
+        <td><%= kudo.title %></td>
+        <td><%= kudo.content %></td>
+        <td><%= kudo.giver&.email %></td>
+        <td><%= kudo.receiver&.email %></td>
```

Comment:
hej tutaj występuje n+1 query
https://semaphoreci.com/blog/2017/08/09/faster-rails-eliminating-n-plus-one-queries.html

Reply:
skąd wiesz, że tutaj jest n+1 query ? doświadczone oko czy może masz jakiś fajny gem do testowania? :) z tego co pamiętam to '&' miało być sposobem na uniknięcie ifa 'if kudo.giver' (zgodnie z radą Oskara). no chyba, ż teraz mam mieć zapiasne kudo.giver.includes(:email) ?

Reply:
co robi operator `&` możesz przeczytać tutaj https://www.visuality.pl/posts/safe-navigation-operator-vs-try-in-rails

ogólnie jest gem który nazywa się Bullet do testowania n+1 query, ale tutaj mam sprawne oko, bo ja rzadko przy code review odpalam aplikacje :)

N+1 jest dlatego że w controlerze ładujesz `Kudo.all` i to jest jedno query do bazy danych (widać to w kosoli serwera), i ptoem dla każdego kudo jakiego wyświetlasz email givera albo receivera to za każdym razem masz kolejne zapytania, więc tutaj to chyba nawet 2n+1

Reply:
a no i problem `n+1` a problem `if kudo.giver` to dwa różne problemy